### PR TITLE
Fix checkout reset persistence errors

### PR DESCRIPTION
**Pull request by mbj**
Extraction from #398 

* Remove Order#ensure_updated_shipments that actually wants to force
  the state machine re-starts without shipments, but does not update
  them directly in any way.
* Create Order#reset_to_cart method that does change the order state to
  cart but will not attempt to #save!
* This will make error visible to the user

### CODE REVIEW

File: core/app/models/spree/order.rb [hunk 1]
```diff
@@ -393,21 +393,17 @@ def create_proposed_shipments
...
-        restart_checkout_flow
-      end
-    end
+    # @return [self]
+    #
+    # @api private
+    def reset_to_cart
+      return self if completed?
 
-    def restart_checkout_flow
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: FYI that `restart_checkout_flow` is referenced in `spree_auth_devise`. I think you could just replace it with `reset_to_cart` though?

Although I think just `#reset_cart` or even `#reset` might be a better name for this method.

File: core/app/models/spree/order.rb [hunk 2]
```diff
@@ -393,21 +393,17 @@ def create_proposed_shipments
...
-      update_attributes(state: 'cart')
+      shipments.destroy_all
+      self.state = 'cart'
+      update!
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: Does `#update!` persist now? I thought we removed that and/or will be removing it.

Comment (mbj) [inline core/app/models/spree/order.rb]: Update for this PR still persists, as it does not include #490, nevertheless as all call sides do persist on their own, this PR can be independent of #490.

File: core/app/models/spree/order.rb (hunk 1 above)
Comment (mbj) [inline core/app/models/spree/order.rb]: Good catch, I'll go with `#reset`. Note that this PR was not `review-ready` yet so I did not finish my passes.

Comment (dkubb) [inline core/app/models/spree/order.rb]: I know, but I figured I'd mention it now to save you headaches later when integrating with `mrh-spree`.

Comment (mbj) [inline core/app/models/spree/order.rb]: k

**Comment by mbj** [comment]
@dkubb This is ready for review again, there is a followup to `spree_auth_devise` that adapts the API change. https://github.com/MountainRoseHerbs/spree_auth_devise/pull/18

File: core/spec/models/spree/order_spec.rb [hunk 3]
```diff
@@ -642,4 +625,62 @@ def assert_expected_order_state
...
+  describe '#reset' do
+    def apply
+      order.reset
+    end
+
+    context 'on non completed order' do
+      subject(:order) do
+        create(
+          :order,
+          state: 'address',
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: The hash values are not aligned here.

File: core/spec/models/spree/order_spec.rb [hunk 4]
```diff
@@ -642,4 +625,62 @@ def assert_expected_order_state
...
+          shipment_total: 10.to_d
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: If you dropped the `.to_d` does it still work?

File: core/spec/models/spree/order_spec.rb [hunk 5]
```diff
@@ -642,4 +625,62 @@ def assert_expected_order_state
       expect(shipment.stock_location).to eql(stock_location)
     end
   end
+
+  describe '#reset' do
+    def apply
+      order.reset
+    end
+
+    context 'on non completed order' do
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: I would say `on incomplete order`.

File: core/spec/models/spree/order_spec.rb (hunk 3 above)
Comment (mbj) [inline core/spec/models/spree/order_spec.rb]: good catch, fixed.

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -60,6 +60,7 @@ steps:
         ../scripts/setup-ci.sh core
         bundle exec mutant --use rspec --require ./spec/dummy/config/environment --since HEAD~1 -- \
           Spree::Order#complete_order \
+          Spree::Order#reset \
```
Comment (dkubb) [inline .buildkite/pipeline.yml]: I think if this were alphabetically ordered it would be swapped with the line below.

Comment (mbj) [inline .buildkite/pipeline.yml]: gixed.

File: core/spec/models/spree/order_spec.rb (hunk 4 above)
Comment (mbj) [inline core/spec/models/spree/order_spec.rb]: yes, at some point I'd like to fix all `decimal` columsn to raise if they are assigned to non `BigDecimal` instances.

Comment (mbj) [inline core/spec/models/spree/order_spec.rb]: fixed.

File: core/spec/models/spree/order_spec.rb [hunk 7]
```diff
@@ -642,4 +625,62 @@ def assert_expected_order_state
...
+    context 'on incomplete order' do
+      subject(:order) do
+        create(
+          :order,
+          state:          'address',
+          shipment_total: 10
+        )
+      end
+
+      let(:shipment) { create(:shipment, order: order)  }
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: There's an extra space before the `}`.

Comment (mbj) [inline core/spec/models/spree/order_spec.rb]: fixed.

**Comment by dkubb** [comment]
@mbj I am good to merge this on green.

**Pull request merged**